framework/src/utils/FunctionMaterialPropertyDescriptor.C

```diff
@@ -211,32 +234,62 @@ FunctionMaterialPropertyDescriptor<is_ad>::printDebug()
 }
 
 template <bool is_ad>
-const GenericMaterialProperty<Real, is_ad> &
-FunctionMaterialPropertyDescriptor<is_ad>::value() const
+GenericReal<is_ad>
+FunctionMaterialPropertyDescriptor<is_ad>::value(unsigned int qp) const
 {
-  if (_value == nullptr)
+  // current property
+  if (_state == PropertyState::CURRENT)
   {
+    if (_value == nullptr)
+    {
+      DerivativeMaterialInterface<Material> * _material_parent =
+          dynamic_cast<DerivativeMaterialInterface<Material> *>(_parent);
+      DerivativeMaterialInterface<Kernel> * _kernel_parent =
+          dynamic_cast<DerivativeMaterialInterface<Kernel> *>(_parent);
+
+      // property name
+      auto name = derivativePropertyName(_base_name, _derivative_symbols);
+
+      // get the material property reference
+      if (_material_parent)
+        _value = _required ? &(_material_parent->getGenericMaterialProperty<Real, is_ad>(name))
+                           : &(_material_parent->getGenericZeroMaterialProperty<Real, is_ad>(name));
+      else if (_kernel_parent)
+        _value = _required ? &(_kernel_parent->getGenericMaterialProperty<Real, is_ad>(name))
+                           : &(_kernel_parent->getGenericZeroMaterialProperty<Real, is_ad>(name));
+      else
+        mooseError("A FunctionMaterialPropertyDescriptor must be owned by either a Material or a "
+                   "Kernel object.");
+    }
+
+    return qp != libMesh::invalid_uint ? (*_value)[qp] : 0.0;
```

Comment:
what s up with the invalid qp value?

Reply:
The system needs to call value once during the construction to ensure that material property dependencies are properly registered. I had to change the signature of the function and the return value, because no I cannot return a reference (as current and old properties have different types when using AD). The new call returns a value at a QP (upcast to AD if `is_ad==true`). However, the property storage is not initialized at that point (`size()==0`) so I cannot access a QP value. Thus for that initialization step we can call `value()` without a QP to merely register the property.